integration/proxy/proxy_test.go

```diff
@@ -754,6 +755,100 @@ func TestALPNSNIProxyDatabaseAccess(t *testing.T) {
 			require.NoError(t, client.Close())
 		})
 	})
+
+	t.Run("authenticated tunnel with cert renewal", func(t *testing.T) {
+		ctx := context.Background()
+		clt := pack.Root.Cluster.Process.GetAuthServer()
+
+		// setup a mock user with a custom role
+		dbUser := "root-db-user"
+		roleName := "session-ttl-limited-access"
+		userName := "alice-cert-tunnel-test"
+		user, err := types.NewUser(userName)
+		require.NoError(t, err)
+		user.AddRole(roleName)
+
+		sessionTTLLimitedRole, err := types.NewRoleV3(roleName, types.RoleSpecV5{
+			Allow: types.RoleConditions{
+				DatabaseLabels: types.Labels{
+					types.Wildcard: []string{types.Wildcard},
+				},
+				DatabaseNames: []string{types.Wildcard},
+				DatabaseUsers: []string{types.Wildcard},
+			},
+			Options: types.RoleOptions{
+				MaxSessionTTL: types.Duration(5 * time.Second),
+			},
+		})
+		require.NoError(t, err)
+		now := time.Now()
+		require.NoError(t, clt.UpsertRole(ctx, sessionTTLLimitedRole))
+		require.NoError(t, clt.UpsertUser(user))
+		// wait for the user and role to be created
+		helpers.WaitForAuditEventTypeWithBackoff(t, clt, now, events.UserCreateEvent)
+		helpers.WaitForAuditEventTypeWithBackoff(t, clt, now, events.RoleCreatedEvent)
+
+		// get a teleport client
+		tc, err := pack.Root.Cluster.NewClient(helpers.ClientConfig{
+			Login:   userName,
+			Cluster: pack.Root.Cluster.Secrets.SiteName,
+		})
+		require.NoError(t, err)
+
+		routeToDatabase := tlsca.RouteToDatabase{
+			ServiceName: pack.Root.MysqlService.Name,
+			Protocol:    pack.Root.MysqlService.Protocol,
+			Username:    dbUser,
+		}
+		m, err := alpnproxy.NewDBCertChecker(tc, routeToDatabase, nil)
+		require.NoError(t, err)
+
+		// configure local proxy without certs but with cert checking/reissuing middleware
+		// local proxy middleware should fetch a DB cert when the local proxy starts
+		lp1 := mustStartALPNLocalProxyWithConfig(t, alpnproxy.LocalProxyConfig{
```

Comment:
This new test might be flaky. Since our CI can be slow, if it takes more than session time before hitting this client call here, the test can be stuck with the prompt (simulated by adding a 6 seconds delay here). Maybe we can look back at the fake clock, and/or maybe use an interface/mock for some part of the middleware?

Reply:
Well, I had hoped to make the integration test actually show cert renewal working instead of mocking middleware logic. I can inject the fake clock into the middleware itself, making it think that cert renewal is required and triggering the cert renewal, but what I didn't like about this is that the test would not fail with bad certs even if middleware did nothing at all - so to make it possible for such a test to fail I would need to examine the local proxy certs before/after each connection and verify that they have been changed.

I'll look at this again

Reply:
Cool. I think I am more concerned about the prompt, as long as it doesn't block when the test fails.

Reply:
Changed the test to just compare certs before/after mocked expiration time instead of using a short TTL cert with real waiting